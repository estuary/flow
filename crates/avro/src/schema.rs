use super::{FLOW_EXTRA_NAME, FLOW_KEY_NAME};
use apache_avro::schema as avro;
use json::schema::formats::Format;
use json::schema::types;
use std::fmt::Write;

/// Map a Shape at the given location into an AVRO schema.
/// If the location is not required and has no default, it may implicitly be none.
pub fn shape_to_avro(loc: json::Location, shape: doc::Shape, required: bool) -> avro::Schema {
    let mut type_ = shape.type_;

    // Is this location nullable ? NULL may union with any other schema.
    let nullable = if shape.type_.overlaps(types::NULL) || (!required && shape.default.is_none()) {
        type_ = type_ - types::NULL;
        true
    } else {
        false
    };

    let base = if type_.overlaps(types::STRING) {
        match (type_ - types::STRING, shape.string.format) {
            // If the location may contain a fractional number, map to a double.
            // This could lose precision of a string number, but we're already using f64's,
            // and these are typically representing Nan or Inf.
            (types::INT_OR_FRAC | types::FRACTIONAL, Some(Format::Number | Format::Integer)) => {
                avro::Schema::Double
            }
            // Or, if the location may contain an integer, map to a string.
            // A string integer may not fit in a 64-bit signed integer.
            (types::INTEGER, Some(Format::Number | Format::Integer)) => avro::Schema::String,
            // String which holds an RFC-3339 date.
            (types::INVALID, Some(Format::Date)) => avro::Schema::Date,
            // String which holds an RFC-3339 date-time.
            (types::INVALID, Some(Format::DateTime)) => avro::Schema::TimestampMicros,
            // String which holds a canonical UUID.
            (types::INVALID, Some(Format::Uuid)) => avro::Schema::Uuid,
            // String which holds an ISO-8601 duration.
            (types::INVALID, Some(Format::Duration)) => avro::Schema::Duration,
            // A regular string.
            (types::INVALID, _) => avro::Schema::String,
            // Other combinations fall back to JSON encoding.
            _ => raw_json_schema(loc),
        }
    } else {
        match type_ {
            types::ARRAY => array_to_avro(loc, shape.array),
            types::BOOLEAN => avro::Schema::Boolean,
            types::INTEGER => avro::Schema::Long,
            types::INT_OR_FRAC | types::FRACTIONAL => avro::Schema::Double,
            types::OBJECT => object_to_avro(loc, shape.object),
            // Other combinations fall back to JSON encoding.
            _ => raw_json_schema(loc),
        }
    };

    if !nullable || matches!(base, avro::Schema::Null) {
        base
    } else {
        avro::Schema::Union(avro::UnionSchema::new(vec![base, avro::Schema::Null]).unwrap())
    }
}

// Map a key extracted from a Shape into a flattened Avro Record.
pub fn key_to_avro(key: &[doc::Pointer], shape: doc::Shape) -> avro::Schema {
    let loc_root = json::Location::Root;
    let loc_key = loc_root.push_prop("Key");
    let loc_parts = loc_key.push_prop("Parts");

    let mut parts = Vec::new();

    // Map each key component into a field with a structured name.
    for (position, ptr) in key.iter().enumerate() {
        let name = format!("p{}", position + 1);
        let (shape, _) = shape.locate(ptr);

        let default = shape.default.as_ref().map(|d| d.0.clone());
        let schema = shape_to_avro(loc_parts.push_prop(&name), shape.clone(), true);

        parts.push(avro::RecordField {
            aliases: None,
            custom_attributes: Default::default(),
            default,
            doc: None,
            name,
            order: avro::RecordFieldOrder::Ascending,
            position,
            schema,
        });
    }

    // Wrap the record of key components in a single-field parent record.
    let parts_schema = avro::Schema::Record(avro::RecordSchema {
        name: location_to_name(loc_parts),
        aliases: None,
        doc: None,
        fields: parts,
        attributes: Default::default(),
        lookup: Default::default(),
    });

    avro::Schema::Record(avro::RecordSchema {
        name: location_to_name(loc_key),
        aliases: None,
        doc: None,
        fields: vec![avro::RecordField {
            aliases: None,
            custom_attributes: Default::default(),
            default: None,
            doc: None,
            name: FLOW_KEY_NAME.to_string(),
            order: avro::RecordFieldOrder::Ascending,
            position: 0,
            schema: parts_schema,
        }],
        attributes: Default::default(),
        lookup: Default::default(),
    })
}

// Map an Object Shape into an Avro schema (a Record or Map).
fn object_to_avro(loc: json::Location, obj: doc::shape::ObjShape) -> avro::Schema {
    let mut fields: Vec<avro::RecordField> = Vec::new();
    let mut extra = doc::Shape::nothing();

    for pattern in obj.pattern_properties {
        extra = doc::Shape::union(extra, pattern.shape);
    }
    if let Some(addl) = obj.additional_properties {
        extra = doc::Shape::union(extra, *addl);
    } else {
        extra = doc::Shape::union(extra, doc::Shape::anything());
    }

    // If there are no explicit properties, but this object may have pattern or
    // additional properties, then interpret it as an Avro map.
    if extra.type_ != types::INVALID && obj.properties.is_empty() {
        let schema = shape_to_avro(loc, extra, true);
        return avro::Schema::map(schema);
    }

    // Otherwise, build a Record which may have a placeholder
    // field for additional dynamic properties.

    for prop in obj.properties {
        if !AVRO_FIELD_RE.is_match(&prop.name) {
            extra = doc::Shape::union(extra, prop.shape);
            continue; // Cannot be represented under Avro's name restrictions.
        }
        let default = prop.shape.default.as_ref().map(|d| d.0.clone());
        let schema = shape_to_avro(loc.push_prop(&prop.name), prop.shape, prop.is_required);

        fields.push(avro::RecordField {
            aliases: None,
            custom_attributes: Default::default(),
            default,
            doc: None,
            name: prop.name.to_string(),
            order: avro::RecordFieldOrder::Ascending,
            position: fields.len(),
            schema,
        })
    }

    // If extra dynamic properties are possible, add a trailing field for them.
    // This field MUST appear last in the record schema.

    if extra.type_ != types::INVALID {
        let schema = shape_to_avro(loc.push_prop(FLOW_EXTRA_NAME), extra, true);
        let schema = avro::Schema::map(schema);

        fields.push(avro::RecordField {
            aliases: None,
            custom_attributes: Default::default(),
            default: None,
            doc: None,
            name: FLOW_EXTRA_NAME.to_string(),
            order: avro::RecordFieldOrder::Ascending,
            position: fields.len(),
            schema,
        })
    }

    avro::Schema::Record(avro::RecordSchema {
        name: location_to_name(loc),
        aliases: None,
        doc: None,
        fields,
        attributes: Default::default(),
        lookup: Default::default(),
    })
}

// Map an Array Shape into an Avro Array.
fn array_to_avro(loc: json::Location, shape: doc::shape::ArrayShape) -> avro::Schema {
    let mut items = doc::Shape::nothing();

    if let Some(addl) = shape.additional_items {
        items = doc::Shape::union(items, *addl);
    } else if !matches!(shape.max_items, Some(m) if m as usize <= shape.tuple.len()) {
        items = doc::Shape::union(items, doc::Shape::anything());
    }

    for shape in shape.tuple {
        items = doc::Shape::union(items, shape);
    }

    let items = shape_to_avro(loc.push_prop("_items"), items, true);
    avro::Schema::array(items)
}

// Map a location into a special Schema holding a string-encoded JSON value.
fn raw_json_schema(loc: json::Location) -> avro::Schema {
    let fields = vec![avro::RecordField {
        aliases: None,
        custom_attributes: Default::default(),
        default: None,
        doc: None,
        name: "json".to_string(),
        order: avro::RecordFieldOrder::Ignore,
        position: 0,
        schema: avro::Schema::String,
    }];

    avro::Schema::Record(avro::RecordSchema {
        name: location_to_name(loc.push_prop("RawJSON")),
        aliases: None,
        doc: None,
        fields,
        attributes: Default::default(),
        lookup: Default::default(),
    })
}

// Map a Location into an Avro dot-separated namespace and name.
fn location_to_name(loc: json::Location) -> avro::Name {
    let name = loc.fold(String::new(), move |loc, mut n: String| {
        match loc {
            json::Location::Root => n.push_str("root"),
            json::Location::Property(json::LocatedProperty { name, .. }) => {
                n.push('.');
                n.push_str(name);
            }
            json::Location::Item(json::LocatedItem { index, .. }) => {
                write!(&mut n, ".{index}").unwrap();
            }
            _ => unreachable!(),
        };
        n
    });

    avro::Name::new(&name).unwrap()
}

lazy_static::lazy_static! {
    // The set of allowed characters in an AVRO field name.
    pub static ref AVRO_FIELD_RE : regex::Regex = regex::Regex::new(r"^[A-Za-z_][A-Za-z0-9_]*$").unwrap();
}

#[cfg(test)]
mod test {
    use serde_json::json;

    #[test]
    fn test_complex_schema() {
        let fixture = json!({
          "type": "object",
          "properties": {
            "a_bool": {"type": "boolean"},
            "a_date": {"type": "string", "format": "date"},
            "a_date_time": {"type": "string", "format": "date-time"},
            "a_duration": {"type": "string", "format": "duration"},
            "a_fractional": {"type": "number", "not": {"type": "integer"}},
            "a_integer_like": {"oneOf": [{"type": "integer"}, {"type": "string", "format": "integer"}]},
            "a_null": {"type": "null"},
            "a_number_like":  {"oneOf": [{"type": "number"}, {"type": "string", "format": "number"}]},
            "a_number_like_mixed1": {"oneOf": [{"type": "integer"}, {"type": "string", "format": "number"}]},
            "a_number_like_mixed2": {"oneOf": [{"type": "number"}, {"type": "string", "format": "integer"}]},
            "a_str": {"type": "string", "default": "xyz"},
            "an_int": {"type": "integer", "default": 32},
            "arr_items": {"type": "array", "items": {"type": "boolean"}},
            "arr_tuple": {
              "type": "array",
              "items": [{"type": "number"}, {"type": "string", "format": "number"}],
              "maxItems": 2,
            },
            "arr_tuple_items": {
              "type": "array",
              "items": [{"type": "integer"}, {"type": "integer"}],
              "additionalItems": {"type": "number"},
            },
            "map": {
                "type": "object",
                "additionalProperties": {"type": "boolean"},
            },
            "obj": {
              "type": "object",
              "properties": {
                "bar": {
                  "type": "object",
                  "properties": {"a_const": {"const": 42}, "filtered invalid avro name": {"const": 52}},
                  "additionalProperties": {"type": "integer"},
                  "patternProperties": {
                    "str_.*": {"type": "string", "format": "integer"},
                  },
                  "required": ["a_const"],
                },
                "foo": {"type": "boolean"},
              },
              "required": ["foo", "bar"],
              "additionalProperties": false,
            },
          },
          "required": ["an_int", "a_bool", "a_date", "obj"],
        }).to_string();

        let key = &["/a_bool", "/obj/a_map/a_const"];
        let key: Vec<_> = key.iter().map(|p| doc::Pointer::from_str(p)).collect();
        insta::assert_json_snapshot!(schema_test(&fixture, &key));
    }

    fn schema_test(json_schema: &str, key: &[doc::Pointer]) -> serde_json::Value {
        let (key, value) = crate::json_schema_to_avro(json_schema, key).unwrap();

        json!({
            "key": &key,
            "value": &value,
        })
    }
}
