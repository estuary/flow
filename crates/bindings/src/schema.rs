use crate::service::{self, Channel};
use derive::schema_api::API;

#[no_mangle]
pub extern "C" fn schema_create(log_level: i32, log_dest_fd: i32) -> *mut Channel {
    service::create::<API>(log_level, log_dest_fd)
}
#[no_mangle]
pub extern "C" fn schema_invoke1(ch: *mut Channel, i: service::In1) {
    service::invoke::<API, _>(ch, i)
}
#[no_mangle]
pub extern "C" fn schema_invoke4(ch: *mut Channel, i: service::In4) {
    service::invoke::<API, _>(ch, i)
}
#[no_mangle]
pub extern "C" fn schema_invoke16(ch: *mut Channel, i: service::In16) {
    service::invoke::<API, _>(ch, i)
}
#[no_mangle]
pub extern "C" fn schema_drop(ch: *mut Channel) {
    service::drop::<API>(ch)
}
