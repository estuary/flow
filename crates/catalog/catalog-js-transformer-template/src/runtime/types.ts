/* eslint-disable @typescript-eslint/no-explicit-any */
export type Document = any;
/* eslint-enable @typescript-eslint/no-explicit-any */

// BootstrapLambda is the type shape of a catalog bootstrap lambda.
export type BootstrapLambda = () => Promise<void>;

// UpdateLambda takes a document and returns an array of register update
// Documents.
export type UpdateLambda = (source: Document) => Promise<Document[]>;

// PublishLambda takes a source document, a previous register, and a
// next register, and returns an array of derived documents to publish.
export type PublishLambda = (source: Document, previous: Document, register: Document) => Promise<Document[]>;

// TransformMap indexes "update" and "publish" lambdas on their catalog
// transform_id.
export interface TransformMap {
    [transformId: number]: {
        update?: UpdateLambda;
        publish?: PublishLambda;
    };
}

// BootstrapMap indexes BootstrapLambdas on their catalog derivation_id.
export interface BootstrapMap {
    [derivationId: number]: BootstrapLambda[];
}
