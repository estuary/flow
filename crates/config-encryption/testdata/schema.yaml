type: object
properties:
  supa_secret:
    secret: true
    type: string
  supda_dupa_secret:
    airbyte_secret: true
    type: integer
  not_secret:
    type: string

