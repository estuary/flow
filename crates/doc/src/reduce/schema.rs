use super::{count_nodes_heap, Cursor, Error, Result};
use crate::{inference::Shape, schema::SchemaBuilder, AsNode, HeapNode};
use json::schema::index::IndexBuilder;

pub fn json_schema_merge<'alloc, L: AsNode, R: AsNode>(
    cur: Cursor<'alloc, '_, '_, '_, '_, L, R>,
) -> Result<HeapNode<'alloc>> {
    let Cursor {
        tape,
        loc,
        full: _,
        lhs,
        rhs,
        alloc,
    } = cur;

    let (lhs, rhs) = (lhs.into_heap_node(alloc), rhs.into_heap_node(alloc));

    *tape = &tape[count_nodes_heap(&rhs)..];

    // Ensure that we're working with objects on both sides
    // Question: Should we actually relax this to support
    // reducing valid schemas like "true" and "false"?
    let (
        lhs @ HeapNode::Object(_),
        rhs @ HeapNode::Object(_)
    ) = (lhs, rhs) else {
        return Err(Error::with_location(Error::JsonSchemaMergeWrongType { detail: None }, loc) )
    };

    let left = shape_from_node(lhs).map_err(|e| Error::with_location(e, loc))?;
    let right = shape_from_node(rhs).map_err(|e| Error::with_location(e, loc))?;

    let mut merged_shape = Shape::union(left, right);
    merged_shape.enforce_field_count_limits(json::Location::Root);

    // Union together the LHS and RHS, and convert back from `Shape` into `HeapNode`.
    let merged_doc = serde_json::to_value(&SchemaBuilder::new(merged_shape).root_schema())
        .and_then(|value| HeapNode::from_serde(value, alloc))
        .map_err(|e| {
            Error::with_location(
                Error::JsonSchemaMergeWrongType {
                    detail: Some(e.to_string()),
                },
                loc,
            )
        })?;

    Ok(merged_doc)
}

fn shape_from_node<'a, N: AsNode>(node: N) -> Result<Shape> {
    // Should this be something more specific/useful?
    let url = url::Url::parse("json-schema-reduction:///").unwrap();

    let serialized =
        serde_json::to_value(node.as_node()).map_err(|e| Error::JsonSchemaMergeWrongType {
            detail: Some(e.to_string()),
        })?;

    let schema = json::schema::build::build_schema::<crate::Annotation>(url.clone(), &serialized)
        .map_err(|e| Error::JsonSchemaMergeWrongType {
        detail: Some(e.to_string()),
    })?;

    let mut index = IndexBuilder::new();
    index.add(&schema).unwrap();
    index.verify_references().unwrap();
    let index = index.into_index();

    Ok(Shape::infer(
        index
            .must_fetch(&url)
            .map_err(|e| Error::JsonSchemaMergeWrongType {
                detail: Some(e.to_string()),
            })?,
        &index,
    ))
}

#[cfg(test)]
mod test {
    use super::super::test::*;
    use super::*;

    #[test]
    fn test_merge_json_schemas() {
        run_reduce_cases(
            json!({ "reduce": { "strategy": "jsonSchemaMerge" } }),
            vec![
                Partial {
                    rhs: json!({
                        "type": "string",
                        "maxLength": 5,
                        "minLength": 5
                    }),
                    expect: Ok(json!({
                        "type": "string",
                        "maxLength": 5,
                        "minLength": 5
                    })),
                },
                Partial {
                    rhs: json!("oops!"),
                    expect: Err(Error::JsonSchemaMergeWrongType { detail: None }),
                },
                Partial {
                    rhs: json!({
                        "type": "foo"
                    }),
                    expect: Err(Error::JsonSchemaMergeWrongType {
                        detail: Some(
                            r#"at keyword 'type' of schema 'json-schema-reduction:///': expected a type or array of types: invalid type name: 'foo'"#.to_owned(),
                        ),
                    }),
                },
                Partial {
                    rhs: json!({
                        "type": "string",
                        "minLength": 8,
                        "maxLength": 10
                    }),
                    expect: Ok(json!({
                        "$schema": "http://json-schema.org/draft-07/schema#",
                        "type": "string",
                        "minLength": 5,
                        "maxLength": 10,
                    })),
                },
            ],
        )
    }
}
