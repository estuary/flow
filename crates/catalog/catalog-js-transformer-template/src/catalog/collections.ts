// This file contains placeholder examples of collection types,
// which are over-written during catalog build.

export type ExampleSourceCollection = {
    hello: string;
};

export type ExampleDerivedCollection = {
    world: string;
    value: number;
};
