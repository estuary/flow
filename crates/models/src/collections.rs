use super::{CompositeKey, Derivation, Field, Id, JournalTemplate, JsonPointer, Schema};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use serde_json::{from_value, json};
use std::collections::BTreeMap;

/// Collection describes a set of related documents, where each adheres to a
/// common schema and grouping key. Collections are append-only: once a document
/// is added to a collection, it is never removed. However, it may be replaced
/// or updated (either in whole, or in part) by a future document sharing its
/// key. Each new document of a given key is "reduced" into existing documents
/// of the key. By default, this reduction is achieved by completely replacing
/// the previous document, but much richer reduction behaviors can be specified
/// through the use of annotated reduction strategies of the collection schema.
#[derive(Serialize, Deserialize, Clone, Debug, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
#[schemars(example = "CollectionDef::example")]
pub struct CollectionDef {
    /// # Schema against which collection documents are validated and reduced on write and read.
    #[schemars(example = "Schema::example_relative")]
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub schema: Option<Schema>,
    /// # Schema against which collection documents are validated and reduced on write.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub write_schema: Option<Schema>,
    /// # Schema against which collection documents are validated and reduced on read.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub read_schema: Option<Schema>,
    /// # Composite key of this collection.
    pub key: CompositeKey,
    /// # Projections and logical partitions of this collection.
    #[schemars(schema_with = "projections_schema")]
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub projections: BTreeMap<Field, Projection>,
    /// # Template for journals of this collection.
    #[serde(default, skip_serializing_if = "JournalTemplate::is_empty")]
    pub journals: JournalTemplate,
    // # Derivation which builds this collection as transformations of other collections.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub derive: Option<Derivation>,
    /// # Expected publication ID of this collection within the control plane.
    /// When present, a publication of the collection will fail if the
    /// last publication ID in the control plane doesn't match this value.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expect_pub_id: Option<Id>,
    /// # Delete this collection within the control plane.
    /// When true, a publication will delete this collection.
    #[serde(default, skip_serializing_if = "super::is_false")]
    pub delete: bool,
}

impl CollectionDef {
    pub fn example() -> Self {
        Self {
            schema: Some(Schema::example_inline_basic()),
            write_schema: None,
            read_schema: None,
            key: CompositeKey::example(),
            projections: BTreeMap::new(),
            journals: JournalTemplate::default(),
            derive: None,
            expect_pub_id: None,
            delete: false,
        }
    }
}

/// Projections are named locations within a collection document which
/// may be used for logical partitioning or directly exposed to databases
/// into which collections are materialized.
#[derive(Serialize, Deserialize, Clone, Debug, JsonSchema)]
#[serde(untagged, deny_unknown_fields, rename_all = "camelCase")]
pub enum Projection {
    Pointer(JsonPointer),
    Extended {
        /// # Location of this projection.
        location: JsonPointer,
        /// # Is this projection a logical partition?
        #[serde(default)]
        partition: bool,
    },
}

impl Projection {
    pub fn as_parts(&self) -> (&JsonPointer, bool) {
        match self {
            Self::Pointer(location) => (location, false),
            Self::Extended {
                location,
                partition,
            } => (location, *partition),
        }
    }

    fn example_pointer() -> Self {
        Self::Pointer(JsonPointer::example())
    }

    fn example_extended() -> Self {
        Projection::Extended {
            location: JsonPointer::example(),
            partition: true,
        }
    }
}

fn projections_schema(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
    let schema = Field::json_schema(gen);
    gen.definitions_mut().insert(Field::schema_name(), schema);

    let schema = Projection::json_schema(gen);
    gen.definitions_mut()
        .insert(Projection::schema_name(), schema);

    from_value(json!({
        "type": "object",
        "patternProperties": {
            Field::schema_pattern(): {
                "$ref": format!("#/definitions/{}", Projection::schema_name()),
            },
        },
        "additionalProperties": false,
        "examples": [{
            "a_field": Projection::example_pointer(),
            "a_partition": Projection::example_extended(),
        }],
    }))
    .unwrap()
}

impl super::ModelDef for CollectionDef {
    fn sources(&self) -> impl Iterator<Item = &crate::Source> {
        self.derive
            .iter()
            .map(|derive| derive.transforms.iter().map(|transform| &transform.source))
            .flatten()
    }
    fn targets(&self) -> impl Iterator<Item = &crate::Collection> {
        std::iter::empty()
    }

    fn catalog_type(&self) -> crate::CatalogType {
        crate::CatalogType::Collection
    }
}
