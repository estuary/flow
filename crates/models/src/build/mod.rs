use crate::{names, tables};
use doc::inference::{Exists, Shape};
use json::schema::types;
use protocol::{flow, protocol as broker};
use serde_json::Value;
use std::collections::BTreeMap;

pub fn inference(shape: &Shape, exists: Exists) -> flow::Inference {
    flow::Inference {
        types: shape.type_.to_vec(),
        must_exist: exists.must(),
        title: shape.title.clone().unwrap_or_default(),
        description: shape.description.clone().unwrap_or_default(),
        string: if shape.type_.overlaps(types::STRING) {
            Some(flow::inference::String {
                content_type: shape.string.content_type.clone().unwrap_or_default(),
                format: shape.string.format.clone().unwrap_or_default(),
                is_base64: shape.string.is_base64.unwrap_or_default(),
                max_length: shape.string.max_length.unwrap_or_default() as u32,
            })
        } else {
            None
        },
    }
}

pub fn collection_spec(
    collection: &tables::Collection,
    projections: Vec<flow::Projection>,
) -> flow::CollectionSpec {
    let tables::Collection {
        collection: name,
        scope: _,
        schema,
        key,
    } = collection;

    let partition_fields = projections
        .iter()
        .filter_map(|p| {
            if p.is_partition_key {
                Some(p.field.clone())
            } else {
                None
            }
        })
        .collect();

    flow::CollectionSpec {
        collection: name.to_string(),
        schema_uri: schema.to_string(),
        key_ptrs: key.iter().map(|p| p.to_string()).collect(),
        projections,
        partition_fields,
        uuid_ptr: collection.uuid_ptr(),
        ack_json_template: serde_json::json!({
                "_meta": {"uuid": "DocUUIDPlaceholder-329Bb50aa48EAa9ef",
                "ack": true,
            } })
        .to_string()
        .into(),
    }
}

pub fn journal_selector(
    collection: &names::Collection,
    selector: &Option<names::PartitionSelector>,
) -> broker::LabelSelector {
    let mut include = vec![broker::Label {
        name: "estuary.dev/collection".to_owned(),
        value: collection.to_string(),
    }];
    let mut exclude = Vec::new();

    if let Some(selector) = &selector {
        push_partitions(&selector.include, &mut include);
        push_partitions(&selector.exclude, &mut exclude);
    }

    // LabelSets must be in sorted order.
    include.sort_by(|l, r| (&l.name, &l.value).cmp(&(&r.name, &r.value)));
    exclude.sort_by(|l, r| (&l.name, &l.value).cmp(&(&r.name, &r.value)));

    broker::LabelSelector {
        include: Some(broker::LabelSet { labels: include }),
        exclude: Some(broker::LabelSet { labels: exclude }),
    }
}

// Flatten partition selector fields into a Vec<Label>.
// JSON strings are percent-encoded but un-quoted.
// Other JSON types map to their literal JSON strings.
// *** This MUST match the Go-side behavior! ***
fn push_partitions(fields: &BTreeMap<String, Vec<Value>>, out: &mut Vec<broker::Label>) {
    for (field, value) in fields {
        for value in value {
            let value = match value {
                Value::String(s) => {
                    percent_encoding::utf8_percent_encode(s, percent_encoding::NON_ALPHANUMERIC)
                        .to_string()
                }
                _ => serde_json::to_string(value).unwrap(),
            };
            out.push(broker::Label {
                name: format!("estuary.dev/field/{}", field),
                value,
            });
        }
    }
}

pub fn capture_spec(
    capture: &tables::Capture,
    target: &tables::BuiltCollection,
    name: &str,
    endpoint_type: flow::EndpointType,
    endpoint_config_json: String,
    endpoint_resource_path: Vec<String>,
) -> flow::CaptureSpec {
    flow::CaptureSpec {
        capture: name.to_owned(),
        collection: Some(target.spec.clone()),
        endpoint_name: capture.endpoint.to_string(),
        endpoint_type: endpoint_type as i32,
        endpoint_config_json,
        endpoint_resource_path,
    }
}

fn lambda_spec(
    lambda: &names::Lambda,
    transform: &tables::Transform,
    suffix: &str,
) -> flow::LambdaSpec {
    match lambda {
        names::Lambda::Typescript => flow::LambdaSpec {
            typescript: format!("/{}/{}", transform.group_name(), suffix),
            ..Default::default()
        },
        names::Lambda::Remote(addr) => flow::LambdaSpec {
            remote: addr.clone(),
            ..Default::default()
        },
    }
}

pub fn transform_spec(
    transform: &tables::Transform,
    source: &tables::Collection,
) -> flow::TransformSpec {
    let tables::Transform {
        scope: _,
        derivation,
        priority,
        publish_lambda,
        read_delay_seconds,
        rollback_on_register_conflict,
        shuffle_hash,
        shuffle_key,
        shuffle_lambda,
        source_collection: _,
        source_partitions,
        source_schema,
        transform: name,
        update_lambda,
    } = &transform;

    let shuffle = flow::Shuffle {
        group_name: transform.group_name(),
        source_collection: source.collection.to_string(),
        source_partitions: Some(journal_selector(&source.collection, source_partitions)),
        source_uuid_ptr: source.uuid_ptr(),
        shuffle_key_ptr: shuffle_key
            .as_ref()
            .unwrap_or(&source.key)
            .iter()
            .map(|p| p.to_string())
            .collect(),
        uses_source_key: shuffle_key.is_none(),
        shuffle_lambda: shuffle_lambda
            .as_ref()
            .map(|l| lambda_spec(&l, transform, "Shuffle")),
        source_schema_uri: source_schema
            .as_ref()
            .map(|s| s.to_string())
            .unwrap_or_else(|| source.schema.to_string()),
        uses_source_schema: source_schema.is_none(),
        filter_r_clocks: update_lambda.is_none(),
        hash: *shuffle_hash as i32,
        read_delay_seconds: read_delay_seconds.unwrap_or(0),
        priority: *priority,
    };

    flow::TransformSpec {
        derivation: derivation.to_string(),
        transform: name.to_string(),
        shuffle: Some(shuffle),
        update_lambda: update_lambda
            .as_ref()
            .map(|l| lambda_spec(l, transform, "Update")),
        publish_lambda: publish_lambda
            .as_ref()
            .map(|l| lambda_spec(l, transform, "Publish")),
        rollback_on_register_conflict: *rollback_on_register_conflict,
    }
}

pub fn derivation_spec(
    derivation: &tables::Derivation,
    collection: &tables::BuiltCollection,
    mut transforms: Vec<flow::TransformSpec>,
) -> flow::DerivationSpec {
    let tables::Derivation {
        scope: _,
        derivation: _,
        register_schema,
        register_initial,
    } = derivation;

    transforms.sort_by(|l, r| l.transform.cmp(&r.transform));

    flow::DerivationSpec {
        collection: Some(collection.spec.clone()),
        transforms,
        register_schema_uri: register_schema.to_string(),
        register_initial_json: register_initial.to_string(),
    }
}

pub fn materialization_name(
    endpoint_name: &str,
    endpoint_resource_path: &[impl AsRef<str>],
) -> String {
    let mut parts = vec![endpoint_name];
    parts.extend(endpoint_resource_path.iter().map(AsRef::as_ref));

    // We must produce a name for this materialization which is suitable for use as a shard ID.
    // That restricts us to unicode letters and numbers, plus the symbols `-_+/.=%`.
    let mut name = String::new();

    for c in parts.join("/").chars() {
        match c {
            // Note that '%' is not included (it must be escaped).
            '-' | '_' | '+' | '/' | '.' | '=' => name.push(c),
            _ if c.is_alphanumeric() => name.push(c),
            c => name.extend(percent_encoding::utf8_percent_encode(
                &c.to_string(),
                percent_encoding::NON_ALPHANUMERIC,
            )),
        }
    }

    name
}

#[cfg(test)]
mod test {
    use super::materialization_name;

    #[test]
    fn test_name_escapes() {
        let out = materialization_name(
            "endpoint name",
            &vec![
                "he!lo৬".to_string(),
                "a/part%".to_string(),
                "_¾the-=res+.".to_string(),
            ],
        );
        assert_eq!(&out, "endpoint%20name/he%21lo৬/a/part%25/_¾the-=res+.");
    }
}

pub fn materialization_spec(
    materialization: &tables::Materialization,
    source: &tables::BuiltCollection,
    name: &str,
    endpoint_type: flow::EndpointType,
    endpoint_config_json: String,
    endpoint_resource_path: Vec<String>,
    fields: flow::FieldSelection,
) -> flow::MaterializationSpec {
    flow::MaterializationSpec {
        materialization: name.to_string(),
        collection: Some(source.spec.clone()),
        endpoint_name: materialization.endpoint.to_string(),
        endpoint_type: endpoint_type as i32,
        endpoint_config_json,
        endpoint_resource_path,
        field_selection: Some(fields),
        shuffle: Some(flow::Shuffle {
            group_name: format!("materialize/{}", name),
            source_collection: source.collection.to_string(),
            // Materializations always read all logical partitions.
            source_partitions: Some(journal_selector(&source.collection, &None)),
            source_uuid_ptr: source.spec.uuid_ptr.clone(),
            // Materializations always group by the collection's key.
            shuffle_key_ptr: source.spec.key_ptrs.clone(),
            uses_source_key: true,
            shuffle_lambda: None,
            source_schema_uri: source.spec.schema_uri.clone(),
            uses_source_schema: true,
            // At all times, a given collection key must be exclusively owned by
            // a single materialization shard. Therefore we can subdivide shards
            // on key.
            filter_r_clocks: false,
            // Deprecated.
            hash: flow::shuffle::Hash::None as i32,
            // Never delay materializations.
            read_delay_seconds: 0,
            // Priority has no meaning since there's just one shuffle
            // (we're not joining across collections as transforms do).
            priority: 0,
        }),
    }
}

pub fn test_step_spec(
    test_step: &tables::TestStep,
    collection: &tables::Collection,
) -> flow::test_spec::Step {
    let tables::TestStep {
        scope,
        collection: _,
        documents,
        partitions,
        step_index,
        step_type,
        test: _,
    } = test_step;

    flow::test_spec::Step {
        step_type: *step_type as i32,
        step_index: *step_index,
        step_scope: scope.to_string(),
        collection: collection.collection.to_string(),
        collection_schema_uri: collection.schema.to_string(),
        collection_key_ptr: collection.key.iter().map(|p| p.to_string()).collect(),
        collection_uuid_ptr: collection.uuid_ptr(),
        docs_json_lines: documents
            .iter()
            .map(|d| d.to_string())
            .collect::<Vec<_>>()
            .join("\n"),
        partitions: Some(journal_selector(&collection.collection, partitions)),
    }
}
