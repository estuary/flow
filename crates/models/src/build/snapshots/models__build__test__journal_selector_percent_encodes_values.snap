---
source: crates/models/src/build/mod.rs
expression: labels

---
LabelSelector {
    include: Some(
        LabelSet {
            labels: [
                Label {
                    name: "estuary.dev/collection",
                    value: "the/collection",
                },
                Label {
                    name: "estuary.dev/field/bar",
                    value: "123",
                },
                Label {
                    name: "estuary.dev/field/bar",
                    value: "true",
                },
                Label {
                    name: "estuary.dev/field/foo",
                    value: "a_whole:nother",
                },
                Label {
                    name: "estuary.dev/field/foo",
                    value: "some%2Fval-ue",
                },
            ],
        },
    ),
    exclude: Some(
        LabelSet {
            labels: [
                Label {
                    name: "estuary.dev/field/foo",
                    value: "no&no@no$yes%28%29%3B",
                },
            ],
        },
    ),
}
